lib/fs.js

```diff
@@ -2014,30 +2014,49 @@ function ReadStream(path, options) {
   this.closed = false;
 
   if (this.start !== undefined) {
-    if (typeof this.start !== 'number' || Number.isNaN(this.start)) {
+    if (typeof this.start !== 'number') {
       throw new ERR_INVALID_ARG_TYPE('start', 'number', this.start);
     }
-    if (this.end === undefined) {
-      this.end = Infinity;
-    } else if (typeof this.end !== 'number' || Number.isNaN(this.end)) {
-      throw new ERR_INVALID_ARG_TYPE('end', 'number', this.end);
+
+    if (Number.isNaN(this.start)) {
+      throw new ERR_OUT_OF_RANGE('start', 'not NaN', this.start);
     }
 
-    if (this.start > this.end) {
-      const errVal = `{start: ${this.start}, end: ${this.end}}`;
-      throw new ERR_OUT_OF_RANGE('start', '<= "end"', errVal);
+    if (this.start < 0) {
+      throw new ERR_OUT_OF_RANGE('start', 'positive', this.start);
+    }
+
+    if (!Number.isSafeInteger(this.start)) {
+      throw new ERR_OUT_OF_RANGE('start', 'safe integer', this.start);
     }
 
     this.pos = this.start;
   }
 
-  // Backwards compatibility: Make sure `end` is a number regardless of `start`.
-  // TODO(addaleax): Make the above typecheck not depend on `start` instead.
-  // (That is a semver-major change).
-  if (typeof this.end !== 'number')
+  if (this.end === undefined) {
     this.end = Infinity;
-  else if (Number.isNaN(this.end))
-    throw new ERR_INVALID_ARG_TYPE('end', 'number', this.end);
+  } else if (this.end !== Infinity) {
+    if (typeof this.end !== 'number') {
+      throw new ERR_INVALID_ARG_TYPE('end', 'number', this.end);
+    }
+
+    if (Number.isNaN(this.end)) {
+      throw new ERR_OUT_OF_RANGE('end', 'not NaN', this.end);
+    }
+
+    if (this.end < 0) {
+      throw new ERR_OUT_OF_RANGE('end', 'positive', this.end);
+    }
+
+    if (!Number.isSafeInteger(this.end)) {
+      throw new ERR_OUT_OF_RANGE('end', 'safe integer', this.end);
```

Comment:
The message will be: `The value of "end" is out of range. It must be safe integer. Received XXX`.
That should be further improved. My suggestion is to do this in two steps:

One would say: `It must be an integer`. If it is not a safe integer, I would say: `It must be => 0 and <= 2 ** 53 - 1`.

So the code block here should look like e.g.:

```js
if (!Number.isSafeInteger(this.end)) {
  if (typeof this.end !== 'number')
    throw new ERR_INVALID_ARG_TYPE('end', 'number', this.end);
  if (!Number.isInteger(this.end))
    throw new ERR_OUT_OF_RANGE('end', 'an integer', this.end);
  throw new ERR_OUT_OF_RANGE('end', '=> 0 and <= 2 ** 53 - 1', this.end);
} else if (this.end < 0) {
  throw new ERR_OUT_OF_RANGE('end', => 0 and <= 2 ** 53 - 1, this.end);
}
if (this.start !== undefined && this.start > this.end) {
  // Code below
}
```

The reason for this code is: a) it is faster because it needs less code path and less checks. b) it unifies the checks to give a consistent error message. c) the error message is clear (NaN will not be >= 0 or <= 2 ** 53 - 1).

To actually realize that the error message is fine or not, it would be nice to also check for the message in the tests.
To do that, I would actually use a loop with values that I want to check for that will all result with the same error message as in:

```js
[
  val1,
  val2,
  val3
].forEach((options) => {
  createReadStreamErr(example, options, rangeError);
});
```

Reply:
@BridgeAR sounds good. Will make changes as required.

Reply:
> `throw new ERR_OUT_OF_RANGE('end', '=> 0 and <= 2 ** 53 - 1', this.end);`

I think you mean `>=`, right? Just wanted to confirm.